### PR TITLE
bitcast packed unions to integers

### PR DESCRIPTION
**Pull request by mattnite**
This was put together to enable some code written for [MicroZig](https://github.com/ZigEmbeddedGroup/microzig).

Question: Do tagless unions have some sort of hidden tag for safety checks?

### CODE REVIEW

File: src/value.zig [hunk 1]
```diff
@@ -1430,6 +1438,20 @@ pub const Value = extern union {
                     }
                 },
             },
+            .Union => switch (ty.containerLayout()) {
+                .Auto => unreachable, // Sema is supposed to have emitted a compile error already
+                .Extern => unreachable, // Handled in non-packed writeToMemory
+                .Packed => {
+                    if (ty.unionTagType() == null)
```
Comment (IntegratedQuantum) [inline src/value.zig:1445]: I think this should be `!= null`

File: src/value.zig [hunk 2]
```diff
@@ -1430,6 +1438,20 @@ pub const Value = extern union {
...
+                        @panic("TODO implement writeToPackedMemory for tagged unions");
```
Comment (Vexu) [inline src/value.zig:1446]: Packed unions cannot be tagged.

**Comment by marler8997** [comment]
> Do tagless unions have some sort of hidden tag for safety checks?

I believe the answer is yes if it's not marked extern (I recall Andrew mentioning this on IRC a while back).

**Comment by mattnite** [comment]
> I believe the answer is yes

Okay that's good to know, we have an mmio abstraction in MicroZig and packed unions in here are providing different ways of interpreting memory layout. Safety tags should cause a compiler error the way they are used, but I think I can just turn off safety checks for a couple code blocks and we'll be good :)

**Comment by andrewrk** [comment]
The new behavior test caused #19384.

**Pull request merged**